miner/worker_test.go

```diff
@@ -509,3 +510,214 @@ func testGetSealingWork(t *testing.T, chainConfig *params.ChainConfig, engine co
+	extra := make([]byte, 32+common.AddressLength+crypto.SignatureLength)
+	copy(extra[32:32+common.AddressLength], testBankAddress[:])
+
+	addressBookAddress := common.HexToAddress("0x1111111111111111111111111111111111111111")
+
+	randomAddress := common.HexToAddress("0x2222222222222222222222222222222222222222")
+	randomAddressHash := common.BytesToHash(common.LeftPadBytes(randomAddress.Bytes(), 32))
+
+	// Priority address (the signer) and another address
+	priorityKey := testBankKey
+	priorityAddress := testBankAddress
+
+	priorityAddressHash := common.BytesToHash(common.LeftPadBytes(priorityAddress.Bytes(), 32))
+
+	otherKey, _ := crypto.GenerateKey()
+	otherAddress := crypto.PubkeyToAddress(otherKey.PublicKey)
+
+	// Construct the genesis with initial allocations, storage, and correct extra data
+	genesis := &core.Genesis{
+		Config:    &config,
+		Timestamp: 1,
+		ExtraData: extra,
+		Alloc: core.GenesisAlloc{
+			addressBookAddress: {
+				Storage: map[common.Hash]common.Hash{
+					common.HexToHash("0x01"): randomAddressHash,
+				},
+			},
+			randomAddress: {
+				Storage: map[common.Hash]common.Hash{
+					common.HexToHash("0x00"): priorityAddressHash,
+				},
+			},
+
+			// Both addresses need balance to pay for gas
+			otherAddress: {
+				Balance: big.NewInt(1e18),
+			},
+			priorityAddress: {
+				Balance: big.NewInt(1e18),
+			},
+		},
+	}
+
+	// Commit the genesis block
+	if _, err := genesis.Commit(db, tdb); err != nil {
+		t.Fatalf("Failed to commit genesis: %v", err)
+	}
+	// Create the blockchain from the genesis
+	blockchain, err := core.NewBlockChain(db, nil, genesis, nil, engine, vm.Config{}, nil, nil)
+	if err != nil {
+		t.Fatalf("Failed to create blockchain: %v", err)
+	}
+	defer blockchain.Stop()
+
+	// Build a backend for the worker based on this blockchain
+	backend := newCustomTestWorkerBackend(t, &config, engine, db, blockchain)
+
+	priorityConfig := &Config{
+		Recommit:                   time.Second,
+		GasCeil:                    params.GenesisGasLimit,
+		AddressBookContractAddress: &addressBookAddress,
+	}
+
+	// Create a worker
+	w := newWorker(priorityConfig, &config, engine, backend, new(event.TypeMux), nil, false)
+	w.setEtherbase(testBankAddress)
+	defer w.close()
+
+	// Subscribe to mined block events
+	sub := w.mux.Subscribe(core.NewMinedBlockEvent{})
+	defer sub.Unsubscribe()
+
+	// Start the mining worker
+	w.start()
+
+	// We will repeat the test logic multiple times.
+	// In each iteration, the non-priority transactions will use a higher gas price.
+	// We still expect the priority transactions to appear first (due to custom logic).
+	const attempts = 5
+	const txCount = 3
+
+	for attempt := 1; attempt <= attempts; attempt++ {
+		// Gas prices: priority < non-priority (reversed from normal scenario)
+		priorityGasPrice := big.NewInt(params.InitialBaseFee * 5)
+		nonPriorityGasPrice := big.NewInt(params.InitialBaseFee * 50)
```

Comment:
this is cool - i also tried changing the nonPriorityGasPrice to be lower than the priorityGasPrice and the test still passes which makes sense, the transactions are ordered in priority **regardless** of gasPrice (which is what we want), though i think it makes most sense to test for this case